### ISSUE TITLE
Consider how to support defining protected methods in the CXX bridge

### ISSUE THREAD

**Issue by ahayzen-kdab** [KDAB/cxx-qt#330]
We need to use `friend` as shown in #329 then this needs some API to determine when to generate eg

```rust
#[cxx_qt::bridge]
mod ffi {
    unsafe extern "C++" {
        #[cxx_qt::protected]
        #[cxx_name = "beginResetModel"]
        fn begin_reset_model(self: Pin<&mut MyObjectQt>);
        #[cxx_qt::protected]
        #[cxx_name = "endResetModel"]
        fn end_reset_model(self: Pin<&mut MyObjectQt>);
    }
}
```

**Comment by ahayzen-kdab** [comment]
Like #/328 this is another task that could be helped by having CXX API. As we need to generate `void cxxbridge1$CustomBaseClass$begin_reset_model(::CustomBaseClass& self) noexcept;` from `fn begin_reset_model(self: Pin<&mut MyObjectQt>);`.

If there was a CXX method to go from the Rust signature to the C++ signature this would help us here.

**Comment by LeonMatthesKDAB** [comment]
## New API suggestion:
``` rust
#[cxx_qt::bridge]
pub mod ffi {

    #[cxx_qt::qobject(base = "QAbstractListModel")]
    pub struct MyObject {}

    impl qobject::MyObject {
        cxx_qt::protected! {
            #[cxx_name="beginResetModel"]
            fn begin_reset_model(self: Pin<&mut Self>);
        }

        #[invokable]
        pub fn my_fun(self: Pin<&mut Self>) {
            self.beginResetModel();
        }
    }
}
```
cxx_qt::protected! would be a new syntax, that needs to be within an `impl qobject::MyObject` block.
That way, we make it syntactically clear that methods defined here should only be available from within qobject::MyObject.
Also, we could assert that a correct `self` type exists, and that it only uses either:
- `self: Pin<&mut Self>`
- `&self` 

`cxx_qt::protected!` would then generate the appropriate `extern "C++"` block, as well as the required "friend" methods.
We should be careful here to, if possible, limit access to the generated `extern "C++"` methods, such that they can only be used from the `qobject` module, or ideally, only from `qobject::MyObject` itself.

**Comment by ahayzen-kdab** [comment]
This looks interesting, can we also use it for public methods ? As this then removes the need for the developer to know about `TQt`.

Eg as discussed

```rust
impl qobject::T {
    cxx_qt::inherit! {
        #[protected] 
        fn x(&self)
        // public
        fn y(&self)
    }
}
```
or
```rust
    impl qobject::MyObject {
        cxx_qt::protected! {
            #[cxx_name="beginResetModel"]
            fn begin_reset_model(self: Pin<&mut Self>);
        }

        cxx_qt::public! {
            fn public_base_method(self: Pin<&mut Self>);
        }
    }
```

**Comment by ahayzen-kdab** [comment]
Also macros are token streams so could have an extern "C++" block, it's more explicit but more verbose

```rust
cxx_qt::inherit! {
   extern "C++" {
       #[protected] 
       fn x(&self);
   }
}
```

Maybe just changing `inherit` to `inherit_cpp` ?

```rust
cxx_qt::inherit_cpp! {
       #[protected] 
       fn x(&self);
}
```

This depends if we ever want anything else in the `inherit` block

**Comment by LeonMatthesKDAB** [comment]
Initial implementation is done in #363, it doesn't include using `friend` though, therefore creating a public wrapper of a previously protected method.
This will be fixed in another PR.

**Comment by ahayzen-kdab** [comment]
Closed via #363
